### DISCUSSION TITLE
Implementing DataItem and DataList

### DISCUSSION THREAD

**Discussion by blaize** [fyne-io/fyne#4020] [Q&A] [answered]
Hello, I have read the book of fyne toolkit and I want to improve my app by using data binding.
The book skip the details of how to implement DataItem or DataList.
For exemple, I have slice of struct containing a name and a state. Can someone please make an exemple of code to implement DataItem, DataList, or whatever to display this list in a widget.ListWithData showing the name in a label and the state in a checkbox ?

I have no idea where to start.

Thanks in advance. Love fyne btw, It's by far the best documented toolkit !

**Comment by andydotxyz** [comment, marked as answer]
Using complex types like this you may be better off not using data binding all the way, it can get a little difficult on list and tree with custom types.

However if you want to explore this further you probably need `binding.BindUntypedList(...)` that should at least get you started - though it may require a fair amount of casts initially before you find a prettier way to wrap it in a custom type for your specific struct.

**Reply by blaize** [reply]
Ok, Thanks for your reply. So it's not recomanded to implement our own type and stick with the basic types.
I will experiment a bit with untyped list.

**Reply by andydotxyz** [reply]
My recommendation was merely to get you started. You can implement your own by extending the existing ones, as I hinted - kind of like how you can extend widgets too.

**Reply by blaize** [reply]
Thanks, I successfuly created a struct with some binding.String inside. Passing the untyped to the list, casting in the DataItem function and bind the widgets for the line with the content of the struct.  It works well, the lines of the list can now be dynamically modified.

**Reply by BillBuilt** [reply]
Could you share your solution? I am having the same problem and just cannot figure out to get it working.

**Reply by andydotxyz** [reply]
I cover custom bindings in https://youtu.be/y0rMjc_GqZU?si=p4tANjQmiDclMWKh if video helps...

**Comment by BillBuilt** [comment]
Attempting to emulate data coming from reading an SQL database. There is an additional step to convert the slice of typed structs returned from the SQL read into to a `[]any` that is not included here, but here's what I came up with - based on the example from the [docs](https://developer.fyne.io/binding/list):

```golang
func main() {
	myApp := app.New()
	myWindow := myApp.NewWindow("List Data")
	myWindow.Resize(fyne.Size{800, 600})

	type user struct {
		ID int
		Name string
		Active bool
	}
	item1 := user{
		ID:     1,
		Name:   "User 1",
		Active: true,
	}
	item2 := user{
		ID:     2,
		Name:   "User 2",
		Active: false,
	}
	item3 := user{
		ID:     3,
		Name:   "User 3",
		Active: true,
	}

	l := []any{item1, item2, item3}

	data := binding.NewUntypedList()
	data.Set(l)

	list := widget.NewListWithData(data,
		func() fyne.CanvasObject {
			return widget.NewLabel("template")
		},
		func(i binding.DataItem, o fyne.CanvasObject) {
			t, _ := i.(binding.Untyped).Get()
			bt := binding.NewString()
			user := t.(user)
			bt.Set(user.Name)
			o.(*widget.Label).Bind(bt)
		})

	add := widget.NewButton("Append", func() {
		val := user{
			ID:     data.Length()+1,
			Name:   fmt.Sprintf("Item %d", data.Length()+1),
			Active: true,
		}
		data.Append(val)
	})
	myWindow.SetContent(container.NewBorder(nil, add, nil, nil, list))
	myWindow.ShowAndRun()
}
```
